### ISSUE TITLE
Conditional local statement

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#6392]
More often than not, I want being able to control `local` statements with conditionals. To be fair, this mostly happens in benchmark code, but at the same time, I think it can have practical use cases, as well. Users may want to do an explicit locality check before entering a performance-critical section of the code, and suppress creating wide pointers if they can determine that everything is local during run time.

Currently, what I end up doing generally is:

```chapel
if condition do
  local do
    someKernel()
else
  someKernel()
```
By itself, this feels clunky. Moreover, distilling `someKernel()` may not be very straightforward/natural.

Instead:
```chapel
local condition do
  someKernel()
```
can be very useful in such cases.

Grammar/syntax can be refined:
```chapel
local if condition do
  someKernel()
```
might be an option. but can be confused with 
```chapel
local do if condition then
  someKernel()
```

Another option may be using `when` instead of `if`:
```chapel
local when condition do
  someKernel()
```

It seems like this should require small modifications in the compiler. If there is enough support, I may take this up as a small parser project.

**Comment by ronawho** [comment]
pinging @benharsh

**Comment by benharsh** [comment]
I generally support this kind of feature. One possible option is to redefine the grammar for ``local`` such that it is similar to ``serial``. ``serial`` currently has the following grammar:
```
TSERIAL expr do_stmt
TSERIAL do_stmt
```

We only have a handful of tests that rely on syntax like ``local foo();``, so it wouldn't be that big of a change. I think we'd prefer the local statement to go away long-term, but this might be a good quality of life improvement in the meantime.

**Comment by e-kayrakli** [comment]
I was wondering if there are any objections to this? Should I go ahead and try to add this to the grammar? (I may be able to work on it as early as this week)

**Comment by benharsh** [comment]
I think @bradcray should have a chance to comment as well.

**Comment by bradcray** [comment]
I don't have any problem adding it if it isn't too much of a distraction and agree that following the pattern set by `serial` is the right way to go (including having the lack of an expression mean `local true`.  If we were wed to the `local` keyword for the long-term, I'd definitely want to take it in this direction.  My only reservation is whether it represents too much wasted work if we manage to retire `local` in the next release cycle or two.  But if it's easy to add and saves some pain and suffering in the meantime, I'd say go for it.

One general question for features like this is whether the boolean expression needs to be a `param` or not.  From context, I'm guessing the proposal here is "not" which is fine with me.  Ideally, if the expression were a `param`, the conditional would get folded down to the appropriate branch (and I could imagine this simply falling out...).

**Comment by e-kayrakli** [comment]
> But if it's easy to add and saves some pain and suffering in the meantime, I'd say go for it.

I have a lightly-tested branch that has this feature working already. I haven't tested it too much, but a simple test and `make check` seems to be working fine with minimal modifications to the internal modules which have stuff like: `local on foo do` (and changed to `local do on foo do`)

> One general question for features like this is whether the boolean expression needs to be a param or not. From context, I'm guessing the proposal here is "not" which is fine with me. Ideally, if the expression were a param, the conditional would get folded down to the appropriate branch (and I could imagine this simply falling out...).

Yes, the answer would be "not". The current implementation I have generates an AST identical to

```chapel
if condition {
  local {
    // stmt
  }
}
else {
  // stmt
}
```
Which leads me to believe that it would follow the familiar semantics of `param` conditionals.

**Comment by e-kayrakli** [comment]
@bradcray -- just wanted to clarify one thing:

> including having the lack of an expression mean local true

In `serial` statement, the conditional added has only one CallExpr to PRIM_SET_SERIAL, so interpreting `serial` as `serial true` has very little impact on the size of the AST. But if we were to interpret `local` as `local true`, that might increase the size of AST more significantly until the `else` block is eliminated during resolution(?).. This sounds like it *might* have an adverse effect on the compilation time and compiler's memory footprint. I don't know enough about the compiler to argue whether this effect would be significant or not. Just wanted to check with you before going any further.

**Comment by bradcray** [comment]
Sounds good to me!

**Comment by bradcray** [comment]
> But if we were to interpret local as local true, that might increase the size of AST more significantly until the else block is eliminated during resolution(?).. 

I was imagining that in interpreting `local` as `local true`, we wouldn't create the whole conditional and have it be folded later, but rather we'd do something more like what I imagine we do today (and simply create the `then` branch if you will).  So I wasn't thinking "implementation approach" so much as "semantics" or even simply "We should be sure to still retain the ability to have just a `local` statement without requiring the conditional expression after it.  Does that help?

**Comment by e-kayrakli** [comment]
> I was imagining that in interpreting local as local true, we wouldn't create the whole conditional and have it be folded later, but rather we'd do something more like what I imagine we do today (and simply create the then branch if you will). So I wasn't thinking "implementation approach" so much as "semantics" or even simply "We should be sure to still retain the ability to have just a local statement without requiring the conditional expression after it. Does that help?

I think so.

Note that the current PR creates both branches for only `local`. What I am understanding is that this was not your concern. So, I should probably change the logic to not create a conditional if the user doesn't specify a conditional.

Piece of grammar that does that (identical to `TSERIAL`):
https://github.com/chapel-lang/chapel/pull/6555/files#diff-03cc4d1786a465ea5afcd52584e69982R484

**Comment by bradcray** [comment]
One thing that occurs to me belatedly with this proposal...   Will the following work?  (and/or, should it?)

```chapel
local {
  localBlahBlahBlah();
  local false {
    remoteBlahBlahBlah();
  }
  localBlahBlahBlah();
}
```

One interpretation would be "the nested local false block will counteract the outer local block, permitting remote accesses."  Another would be "the outer local block asserts that everything within it is local, so the presence of a remote access is a bug, even within a nested local false block."

This example could easily be rewritten as:

```chapel
local {
  localBlahBlahBlah();
}
remoteBlahBlahBlah();
local {
  localBlahBlahBlah();
}
```

but more generally, the code could span multiple dynamic scopes making such cases harder to notice or rewrite.

**Comment by benharsh** [comment]
I do not think that the ``local false`` should override the outer ``local true``, meaning a remote access would be an execution-time error. What if I had such a ``local false`` within a function in some library? It seems weird to me that some arbitrary code could silently avoid the user's assertion that everything is local.

If someone desires this kind of functionality, I think it should be expressed via data-centric features instead of the local block.

**Comment by e-kayrakli** [comment]
> It seems weird to me that some arbitrary code could silently avoid the user's assertion that everything is local.

I agree with this. I think that outer local blocks should override inner ones. However, this sort of brings up the AST bloating question I had before in a different form: even if the inner conditional locals would be overridden by an outer one, inner conditional local would still have both branches. And these would be identical branches, as the else block would be localized by the outer one. But maybe this is not a big issue as you can blow up the AST in easier ways (i.e. big param loops)

Regardless, I understand that this can be confusing, then the question is whether this confusion is small enough to be cleared out by the language specs.

**Comment by bradcray** [comment]
I don't feel strongly about this, particularly since I consider `local` to be a feature that's going away -- I just think it should be tested to make sure it works as expected, and probably documented if/when the documentation changes to support the change in `local` syntax.

It's occurring to me belatedly that this will be a breaking change for current programs, is that right?  I.e., if I don't use the curly brace form, I'll now get a syntax error?

**Comment by e-kayrakli** [comment]
> It's occurring to me belatedly that this will be a breaking change for current programs, is that right? I.e., if I don't use the curly brace form, I'll now get a syntax error?

Correct.

**Comment by bradcray** [comment]
We may want to send a mail to chapel-developers just so nobody is caught by surprise by this then and make sure nobody squawks.  It makes me a little nervous, but at the same time I think that (a) the number of people who use local blocks is low-ish, and (b) those that do probably use the curly bracket form so will be OK.

**Comment by bradcray** [comment]
Oh, and `local` actually never made it into the spec, did it?  That's reassuring as well.

**Comment by e-kayrakli** [comment]
Just realized that this issue is still open. This feature has been merged to master and I think this can be closed (didn't want to do it myself, maybe there is a reason it is open?)

**Comment by benharsh** [comment]
This issue was resolved by PR #6555
